### PR TITLE
[tests-only] Adds e2e tests for download files batch action

### PR DESCRIPTION
**Pull request by SwikritiT**
## Description

Adds test implementation for step `Alice downloads 2 files as tar.`

## Related Issue

- Part of: https://github.com/owncloud/web/issues/6389

## How Has This Been Tested?

- local
- CI

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [x] e2e tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:

- [ ] ...

### CODE REVIEW

**Comment by ownclouders** [comment]
Results for <strong>e2e-tests oC10</strong> https://drone.owncloud.com/owncloud/web/23335/10/1
<details><summary>:boom: To see the trace, please open the link in the console ...</summary>

<p>


#### npx playwright show-trace https://cache.owncloud.com/public/owncloud/web/23335/tracing/alice-creates-public-link-to-a-folder-alice-2022-3-7-09-54-44.zip 


</p></details>

**Comment by SwikritiT** [comment]
The base branch needs to be merged first https://github.com/owncloud/web/pull/6461

File: tests/e2e/support/page/files/allFiles.ts
Comment (amrita-shrestha) [inline tests/e2e/support/page/files/allFiles.ts:5]:
```suggestion
import util from 'util'
```

File: tests/e2e/cucumber/step_definitions/app.files.ts
Comment (SwikritiT) [inline tests/e2e/cucumber/step_definitions/app.files.ts:188]: As suggested in https://github.com/owncloud/web/issues/6389#issuecomment-1060580072 I've skipped this is OC10 and put a comment. Need suggestion on if this approach is okay or we want to write separate tests according to the behavior of diff servers. In this case, writing separate tests means not using this step for `OC10` as other steps will be the same, which is almost equivalent to skipping it here. 

cc: @individual-it @pascalwengerter @fschade @kulmann

File: tests/e2e/cucumber/publicLinkToFolderJourney.feature
Comment (amrita-shrestha) [inline tests/e2e/cucumber/publicLinkToFolderJourney.feature:42]: how about putting comment
```suggestion
      | textfile.txt | folderPublic |
    # Then what do we check for to be confident that the above things done by Brian have worked?
    # Then the downloaded files should have content "abc..."
```

File: tests/e2e/support/page/files/allFiles.ts
Comment (SwikritiT) [reply tests/e2e/support/page/files/allFiles.ts:6]: I'm sorry what is the suggestion here?

File: tests/e2e/cucumber/publicLinkToFolderJourney.feature
Comment (SwikritiT) [reply tests/e2e/cucumber/publicLinkToFolderJourney.feature:42]: The downloaded file has the name `download.tar` and the assertion is done here https://github.com/owncloud/web/blob/6617eae4dd916d38004256ba7c3a00a8ebb0031e/tests/e2e/cucumber/step_definitions/app.files.ts#L200.
I don't know what do you mean by that. You want me to implement 
`Then the downloaded files should have content "ABC..." ` ?

File: tests/e2e/support/page/files/allFiles.ts
Comment (amrita-shrestha) [reply tests/e2e/support/page/files/allFiles.ts:6]: I think we put imported libraries at the top.

File: tests/e2e/cucumber/publicLinkToFolderJourney.feature
Comment (amrita-shrestha) [reply tests/e2e/cucumber/publicLinkToFolderJourney.feature:42]: I think we implement  expect(download.suggestedFilename()).toBe('download.tar')  in then step

File: tests/e2e/support/page/files/allFiles.ts
Comment (saw-jan) [inline tests/e2e/support/page/files/allFiles.ts:153]: I think you don't need this check. As per the playwright doc `page.check()` asserts internally

File: tests/e2e/support/page/files/allFiles.ts
Comment (saw-jan) [inline tests/e2e/support/page/files/allFiles.ts:149]: this way you can make optional param
```suggestion
  async toggleFileOrFolderCheckbox(page, path, checked=true) {
```

**Review by pascalwengerter — CHANGES_REQUESTED** [review]
Please don't merge before @fschade has brought in his E2E improvements

File: tests/e2e/cucumber/step_definitions/app.files.ts
Comment (individual-it) [reply tests/e2e/cucumber/step_definitions/app.files.ts:188]: I would not hide those "workarounds" in the test code. With that the step does not do what it promises.
If this feature does not work in oc10 then I think we need to have an extra scenario for oc10 not using this step

**Comment by SwikritiT** [comment]
> Please don't merge before @fschade has brought in his E2E improvements

@pascalwengerter  Is it limited to this PR only?  There is another PR too for the e2e tests https://github.com/owncloud/web/pull/6461 that's ready to be merged. So let us know if we need to withhold that too.
Also, just out of curiosity what sort of improvements are we talking about?

**Comment by ownclouders** [comment]
Results for <strong>oC10IntegrationApp1</strong> https://drone.owncloud.com/owncloud/web/23909/71/1
<details><summary>:boom: The acceptance tests failed on retry. Please find the screenshots inside ...</summary>

<p>


### webUIPrivateLinks-accessingPrivateLinks_feature-L17.png
  


</p></details>
:boom: The acceptance tests pipeline failed. The build has been cancelled.

File: tests/e2e/cucumber/publicLinkToFolderJourney.feature
Comment (SwikritiT) [reply tests/e2e/cucumber/publicLinkToFolderJourney.feature]: In e2e we are asserting in the same step and not using a lot of then step unless absolutely necessary. Look here
https://github.com/owncloud/web/blob/0984f76d430902ec5011b135b17ae0dabc9f0092/tests/e2e/cucumber/steps/app-files/resource.ts#L58-L60

File: tests/e2e/cucumber/features/integrations/link.feature
Comment (ScharfViktor) [inline tests/e2e/cucumber/features/integrations/link.feature:36]: extra step can be removed with step 21 line

**Review by ScharfViktor — COMMENTED** [review]
few remarks. 
I tested this. I was worried that the check would not work when I downloaded the tar again(download(1).tar) but it works. cool)
I suggest only adding a check that the files are more than 1, because if it is 1 then the tar will not download. It will be download txt file
```
And "Alice" downloads the following files using batch action
      | resource     | from         |
      | lorem.txt    | folderPublic |
```

File: tests/e2e/cucumber/features/integrations/link.feature
Comment (fschade) [reply tests/e2e/cucumber/features/integrations/link.feature:36]: as victor said, we do not have to log out alice in line 21 and as a result no re-login is needed

File: tests/e2e/support/objects/app-files/resource/actions.ts
Comment (fschade) [inline tests/e2e/support/objects/app-files/resource/actions.ts:132]: can we please refactor that part out into a dedicated function to select multiple resources

File: tests/e2e/support/objects/app-files/resource/index.ts
Comment (fschade) [inline tests/e2e/support/objects/app-files/resource/index.ts:50]: can't we just use the download action for that?
Then we have to extend it to support `via: quickaction | sidebar`.

File: tests/e2e/cucumber/steps/app-files/resource.ts
Comment (fschade) [inline tests/e2e/cucumber/steps/app-files/resource.ts:68]: please use the already existing download step for that and extend it with an `via` instruction or use the table steps for that.

File: tests/e2e/cucumber/features/integrations/link.feature
Comment (fschade) [inline tests/e2e/cucumber/features/integrations/link.feature:3]: we've discussed that today and decided to not introduce test tags for now at all.
The idea was to have 3 types of feature files:

* link.feature
* link.ocis.feature
* link.oc10.feature

then wa're able to decide on the the cli level which tests to run by using an glob.

* ocis only -> any-path/*{!.oc10}.feature
* oc10 only -> any-path/*{!.ocis}.feature

please have a look and try it out if it makes any sense for you.

**Review by fschade — CHANGES_REQUESTED** [review]
please see the individual comments, 
thanks for your contribution!

File: tests/e2e/cucumber/features/integrations/link.feature
Comment (fschade) [reply tests/e2e/cucumber/features/integrations/link.feature]: This is implemented now in latest master

**Comment by fschade** [comment]
is this still active or should we take over?

**Comment by SwikritiT** [comment]
> is this still active or should we take over?
I was actually waiting for you to review it again. I have addressed your previous reviews. Can you take a look at it again?

File: tests/e2e/support/objects/app-files/resource/actions.ts
Comment (fschade) [inline tests/e2e/support/objects/app-files/resource/actions.ts:155]: the name `Resources` already indicates that we have multiple resources, can we name it to `selectResources`?

File: tests/e2e/cucumber/steps/app-files/resource.ts
Comment (fschade) [inline tests/e2e/cucumber/steps/app-files/resource.ts:38]: can we please use `resource` instead of `file` here, file isn't a good match if the user tries to download a folder.

**Review by fschade — CHANGES_REQUESTED** [review]
Thanks for the PR, just some minor nitpicks.

File: tests/e2e/support/objects/app-files/resource/actions.ts
Comment (SwikritiT) [reply tests/e2e/support/objects/app-files/resource/actions.ts:17]: Some time ago this step failed most of the time in CI and a few times locally because it won't get `textfile.txt` that was drop uploaded. Locally upon reloading the page the file would appear, couldn't reproduce it manually so I thought it was test timing problem so I added that line to be sure that the selector is preset before we do the click action.  
https://github.com/owncloud/web/blob/b4e626f293826099c284d6612716488a9cf6fe40/tests/e2e/cucumber/features/integrations/link.ocis.feature#L34-L37 

I haven't seen that step fail since then I'll remove it and see if it has the same behavior.

**Comment by SwikritiT** [comment]
@fschade I've addressed your reviews can you take a look?

File: tests/e2e/cucumber/features/integrations/link.oc10.feature
Comment (pascalwengerter) [inline tests/e2e/cucumber/features/integrations/link.oc10.feature:3]: The linked issue has been resolved so after this PR has been merged we should be able to run the directory download tests for both backends (= remove the `oc10` from the filename)

**Review by pascalwengerter — APPROVED** [review]
LGTM, thanks for pushing through here!

**Comment by pascalwengerter** [comment]
Also please use the squash-merge option here:)

**Review by fschade — APPROVED** [review]
LGTM 🎸

**Pull request merged**
